src/schemathesis/specs/openapi/stateful.py

```diff
@@ -0,0 +1,124 @@
+
+if TYPE_CHECKING:
+    from ...models import Case, Endpoint
+    from .schemas import BaseOpenAPISchema
+
+
+def make_new_case(response: requests.Response, case_strategy: SearchStrategy["Case"]) -> st.SearchStrategy["Case"]:
+    @st.composite  # type: ignore
+    def combined_case(draw: st.DataObject) -> "Case":
+        case = draw(case_strategy)
+        case.path_parameters["user_id"] = response.json()["id"]
+        return case
+
+    return combined_case()
+
+
+def call_and_validate(self: RuleBasedStateMachine, case: "Case") -> requests.Response:
+    response = case.call()
+    case.validate_response(response)
+    return response
+
+
+def create_state_machine(schema: "BaseOpenAPISchema", links_field: str) -> Type[RuleBasedStateMachine]:
+    bundles = {}
+    sources: Dict[str, Dict[str, List[str]]] = defaultdict(lambda: defaultdict(list))
+    for endpoint in schema.get_all_endpoints():
+        name = endpoint.verbose_name
+        bundles[name] = Bundle(name)
+        for status_code, definition in endpoint.definition.resolved["responses"].items():
+            for link in definition.get(links_field, {}).values():
+                if "operationId" in link:
+                    target_endpoint = schema.get_endpoint_by_operation_id(link["operationId"])  # type: ignore
+                else:
+                    target_endpoint = schema.get_endpoint_by_reference(link["operationRef"])  # type: ignore
+                target_name = target_endpoint.verbose_name
+                sources[target_name][name].append(status_code)
+
+    rules: Dict[str, Rule] = {}
+
+    for endpoint in schema.get_all_endpoints():
+        make_rule(endpoint, sources, bundles, rules)
+
+    return type("StatefulTest", (RuleBasedStateMachine,), {**bundles, **rules})
+
+
+def make_rule(
+    endpoint: "Endpoint", sources: Dict[str, Dict[str, List[str]]], bundles: Dict[str, Bundle], rules: Dict[str, Rule]
+) -> None:
+    name = endpoint.verbose_name
+    endpoint_strategy = endpoint.as_strategy()
+    if name in sources:
+        strategies = [
+            create_strategy(bundles[source_name], status_code, status_codes, endpoint_strategy)
+            for source_name, status_codes in sources[name].items()
+            for status_code in status_codes
+        ]
+        strategy = functools.reduce(operator.or_, strategies[1:], strategies[0])
+    else:
+        strategy = endpoint_strategy
+        rules[f"INIT_{name}"] = initialize(target=bundles[name], case=endpoint_strategy)(call_and_validate)
+    rules[f"RULE_{name}"] = rule(target=bundles[name], case=strategy)(call_and_validate)
+
+
+def create_strategy(
+    bundle: Bundle, status_code: str, status_codes: List[str], target_strategy: SearchStrategy["Case"]
+) -> SearchStrategy["Case"]:
+    """Create a Hypothesis strategy for an endpoint.
+
+    The cases are generated based on responses that are considered "sources" for this endpoint.
+    """
+    if status_code == "default":
```

Comment:
It would be better if this part will be unaware of that specific filtering logic - it could be passed as a dependency. This way we can decide how to take responses upfront - e.g. links, inference, randomly, etc.